docs/transaction.md

```diff
@@ -0,0 +1,243 @@
+
+### BLPOP and transactions
+If multiple elements are pushed either via a transaction or via variadic arguments of LPUSH command then BLPOP is waked after that transaction or command completely finished. Specifically, when a client performs
+`LPUSH listkey a b c`, then `BLPOP listkey 0` will pop `c`, because `lpush` pushes first `a`, then `b` and then `c` which will be the first one on the left.
+
+If a client executes a transaction that first pushes into a list and then pops from it atomically, then another client blocked on `BLPOP` won’t pop anything, because it waits for the transaction to finish. When BLPOP itself is run in a transaction its blocking behavior is disabled and it returns the “timed-out” response if there is no element to pop.
+
+### Complexity of implementing BLPOP in Dragonfly
+The ordering semantics of BLPOP assume total order of the underlying operations. BLPOP must “observe” multiple keys simultaneously in order to determine which one is non-empty in left-to-right order. If there are no keys with lists, BLPOP blocks, waits, and “observes” which key is being filled first.
+
+For the single-threaded Redis the order is determined by following the natural execution of operations inside the main execution thread.  However, for a multi-threaded, shared-nothing execution, there is no concept of total order or a global synchronized timeline. For non-blockign scenario, "observing" keys is atomic because we lock the keys when executing a command in Dragonfly.
+
+However with blocking scenario for BLPOP, we do not have a built-in mechanism to determine which key was filled earlier - since as I said the concept of total order does not exist for multiple shards.
+
+### Interesing examples to consider:
+
+**Ex1:**
+```
+client1: blpop X, Y  // blocks
+client2: lpush X A
+client3: exist X Y
+```
+
+Client3 should always return 0.
+
+**Ex2:**
+
+```
+client1: BLPOP X Y Z
+client2: RPUSH X A
+client3: RPUSH X B;  RPUSH Y B
+```
+
+**Ex3:**
+
+```
+client1: BLPOP X Y Z
+client2: RPUSH Z C
+client3: RPUSH X A
+client4: RPUSH X B; RPUSH Y B
+```
+
+### BLPOP Ramblings
+There are two cases of how a key can appear and wake a blocking `BLPOP`:
+
+a. with lpush/rpush/rename commands.
+b. via multi-transaction.
+
+`(a)` is actually easy to reason about, because those commands operate on a single key and single key operations are strictly serializable in shared-nothing architecture.
+
+With `(b)` we need to consider the case where we have "BLPOP X Y 0" and then a multi-transaction fills both `y` and `x` using multiple "lpush" commands. Luckily, a multi-transaction in Dragonfly introduces a global barrier across all its shards, and it does not allow any other transactions to run as long as it does not finish. So the blocking "blpop" won't be awaken until the multi-transaction won't finish its run. By that time the state of the keys will be well defined and "blpop" will be able to choose the first non empty key to pop from.
```

Comment:
until the multi-transaction won't finish its run -> until the multi-transaction finishes running 
(Double negation)